lib/Transforms/IPO/StructFieldCacheAnalysis.cpp

```diff
@@ -36,30 +46,672 @@ class StructFieldCacheAnalysisPass : public ModulePass {
+
+  /// Record all users of a GEP instruction/operator that calculates the address of a field.
+  void addFieldAccessFromGEP(const User* U);
+
+  /// Record an access pattern in the data structure for a load/store
+  void addFieldAccessNum(const Instruction* I, FieldNumType FieldNum);
+};
+
+/// This class is inherited from AssemblyAnnotationWriter and used
+/// to print annotated information on IR
+class StructFieldCacheAnalysisAnnotatedWriter : public AssemblyAnnotationWriter {
+ public:
+  StructFieldCacheAnalysisAnnotatedWriter(const StructFieldAccessManager* S = NULL): StructManager(S) {}
+
+  // Override the base class function to print an annotate message after each basic block
+  virtual void emitBasicBlockEndAnnot(const BasicBlock* BB,
+                                      formatted_raw_ostream &OS) {
+    OS.resetColor();
+    auto count = StructManager->getExecutionCount(BB);
+    if (count.hasValue()){
+      OS.changeColor(raw_ostream::YELLOW, false, false);
+      OS << "; [prof count = " << count.getValue() << "]\n";
+      OS.resetColor();
+    }
+    else{
+      OS.changeColor(raw_ostream::YELLOW, false, false);
+      OS << "; [prof count not found " << "]\n";
+      OS.resetColor();
+    }
+  }
+
+  virtual void emitInstructionAnnot(const Instruction *I,
+                                    formatted_raw_ostream &OS) {
+    if (StructManager == NULL)
+      return;
+    if (auto pair = StructManager->getFieldAccessOnInstruction(I)){
+      OS.changeColor(raw_ostream::GREEN, false, false);
+      auto* type = pair.getValue().first;
+      assert(isa<StructType>(type));
+      if (dyn_cast<StructType>(type)->isLiteral())
+        OS << "; [Field " << pair.getValue().second << " of a literal struct.] ";
+      else
+        OS << "; [Field " << pair.getValue().second << " of struct " << type->getStructName() << "] ";
+    }
+    else{
+      OS.resetColor();
+    }
+  }
+ private:
+  const StructFieldAccessManager* StructManager;
+};
+}
+
+// Functions for StructFieldAccessInfo
+void StructFieldAccessInfo::addFieldAccessNum(const Instruction* I, FieldNumType FieldNum)
+{
+  assert(I->getOpcode() == Instruction::Load || I->getOpcode() == Instruction::Store); // Only loads and stores
+  assert (LoadStoreFieldAccessMap.find(I) == LoadStoreFieldAccessMap.end());
+  LoadStoreFieldAccessMap[I] = FieldNum;
+}
+
+Optional<FieldNumType> StructFieldAccessInfo::getAccessFieldNum(const Instruction* I) const
+{
+  Optional<FieldNumType> ret;
+  auto it = LoadStoreFieldAccessMap.find(I);
+  if (it != LoadStoreFieldAccessMap.end()){
+    ret = it->second;
+  }
+  return ret;
+}
+
+FieldNumType StructFieldAccessInfo::calculateFieldNumFromGEP(const User* U) const
+{
+  DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Calculating field number from GEP: " << *U << "\n");
+  //Operand 0 should be a pointer to the struct
+  assert(isa<GetElementPtrInst>(U) || isa<GEPOperator>(U));
+  auto* Op = U->getOperand(0);
```

Comment:
I didn't use that is because U can be either a GetElementPtrInst or a GEPOperator. It needs extra code to handle the two difference cases. Please let me know if I still need to change it.